express-notes-api/server.js

```diff
@@ -0,0 +1,184 @@
+import express from 'express';
+import * as fs from 'node:fs/promises';
+
+/* This just makes it easier to create error messages for me.
+   And it looks cleaner when written in the code later on. */
+class Error {
+  constructor(errorMsg) {
+    this.error = errorMsg;
+  }
+}
+
+const port = 8080;
+const readFile = './data.json';
+const writeFile = './data.json';
+const app = express();
+const parser = express.json();
+app.use(parser);
+
+/* An async function for grabbing the data from the json file.
+   Data will be parsed before returning.                    */
+async function getJSON(fileUrl) {
+  try {
+    const contents = await fs.readFile(fileUrl, 'utf8');
+    const parsed = JSON.parse(contents);
+    return parsed;
+  } catch (err) {
+    console.log('Error reading json:', err);
+  }
+}
+
+/* An async function for writing our updated data back to the json file. */
+async function writeJSON(fileUrl, data) {
+  try {
+    await fs.writeFile(fileUrl, JSON.stringify(data));
+  } catch (err) {
+    console.log('Error writing json back to file: ', err);
+  }
+}
+
+/* For when the user sends a request just to /api/notes, meaning they
+   want all the notes printed, not one specifically               */
+app.get('/api/notes', async (req, res) => {
+  try {
+    const jsonData = await getJSON(readFile);
+    const notesArray = [];
+    for (const grade in jsonData.notes) {
+      notesArray.push(jsonData.notes[grade]);
+    }
+    res.status(200).json(notesArray);
+  } catch (err) {
+    res.status(400).send('Error: could not read notes.');
+    // For the developer to see an error trace, not the user
+    console.log('Error reading notes: ', err);
+  }
+});
+
+/* For when the user sends a request just to /api/notes/:id, meaning
+   they want a specific note printed, and only that note          */
+app.get('/api/notes/:id', async (req, res) => {
+  try {
+    const jsonData = await getJSON(readFile);
+    const requestedId = req.params.id;
```

Comment:
You convert this to Number a few times. Might be easier to convert it once, here.